dll/win32/kernel32/winnls/string/nls.c

```diff
@@ -2237,13 +2272,339 @@ IsDBCSLeadByte(BYTE TestByte)
+        goto Quit;
+    }
+
+    /* Set the DACL to descriptor */
+    Status = RtlSetDaclSecurityDescriptor(&AbsoluteSd,
+                                          TRUE,
+                                          Dacl,
+                                          FALSE);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("CreateNlsDirectorySecurity(): Failed to insert DACL into the descriptor (Status 0x%08x)\n", Status);
+        goto Quit;
+    }
+
+    /* Determine how much size is needed to convert the absolute SD into self-relative one */
+    Status = RtlAbsoluteToSelfRelativeSD(&AbsoluteSd,
+                                         NULL,
+                                         &RelSdSize);
+    if (Status != STATUS_BUFFER_TOO_SMALL)
+    {
+        DPRINT1("CreateNlsDirectorySecurity(): Unexpected status code, must be STATUS_BUFFER_TOO_SMALL (Status 0x%08x)\n", Status);
+        goto Quit;
+    }
+
+    /* Allocate buffer for relative SD */
+    RelativeSd = RtlAllocateHeap(RtlGetProcessHeap(), HEAP_ZERO_MEMORY, RelSdSize);
+    if (RelativeSd == NULL)
+    {
+        DPRINT1("CreateNlsDirectorySecurity(): Could not allocate memory for relative SD, not enough memory!\n");
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto Quit;
+    }
+
+    /* Convert it now */
+    Status = RtlAbsoluteToSelfRelativeSD(&AbsoluteSd,
+                                         RelativeSd,
+                                         &RelSdSize);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("CreateNlsDirectorySecurity(): Failed to convert absolute SD to self-relative format (Status 0x%08x)\n", Status);
+        goto Quit;
+    }
+
+    /* Give the security descriptor to the caller */
+    *NlsSecurityDescriptor = RelativeSd;
+
+Quit:
+    if (WorldSid != NULL)
+    {
+        RtlFreeHeap(RtlGetProcessHeap(), 0, WorldSid);
+    }
+
+    if (AdminsSid != NULL)
+    {
+        RtlFreeHeap(RtlGetProcessHeap(), 0, AdminsSid);
+    }
+
+    if (Dacl != NULL)
+    {
+        RtlFreeHeap(RtlGetProcessHeap(), 0, Dacl);
+    }
+
+    if (!NT_SUCCESS(Status))
+    {
+        if (RelativeSd != NULL)
+        {
+            RtlFreeHeap(RtlGetProcessHeap(), 0, RelativeSd);
+        }
+    }
+
+    return Status;
+}
+
+/**
+ * @brief
+ * Creates a security descriptor for each NLS section
+ * name.
+ *
+ * @param[in] AccessMask
+ * An access mask bit to supply to the function. This
+ * access mask grants everyone access specific to that
+ * bit mask.
+ *
+ * @param[out] SecurityDescriptor
+ * A pointer to an allocated and created security descriptor
+ * that is given to the caller.
+ *
+ * @return
+ * STATUS_SUCCESS is returned if the function has successfully
+ * created a security descriptor for a NLS section name. Otherwise
+ * a NTSTATUS failure code is returned.
+ *
+ * @remarks
+ * The way we implement the function diverges with the Windows' one.
+ * On Windows Server 2003 this function takes 3 parameters. The first one
```

Comment:
And what is the problem of just keeping the 3 existing parameters? First one was your SecurityDescriptor, second one was your AccessMask, and you wouldn't need to do anything with the 3rd one.